model/group.cc

```diff
@@ -385,6 +385,31 @@ namespace minsky
          source.clear();
        }
   }
+  
+  // Move contents out of source group but leave them in selection. Fix for tickets 1080
+  void Group::moveContentsInSelection(Group& source) {
```

Comment:
Why do we need this method? Surely it is enough to loop over the temporary pasted group's contents call ensureItemInserted on that and then addItem on model.

Reply:
When I set "auto p = model", and move g's contents to p, all the items on the canvas, not only the temporary pasted group's, are looped over on using "for (auto i:p->items)" to put the items in selection. I don't know why if only g's contents get moved to p, all items are selected when I loop over those in p? This is what caused the bug you caught and pointed out on sourceforge. So, the only way I could ensure that only the temporary group's contents are looped over and remain in selection, was to do so inside Group::moveContents(). Obviously, when we ungroup items, we don't necessarily want them to be in selection upon ungrouping, so I added a second method to deal with the case where they need to remain in selection.

Reply:
OK - but its a lot of extra code that does stuff, then undoes most of the stuff to leave what you want done. What you want done is two things:

1) insertion of all items and groups of the temporary group into canvas.model (which is not necessarily the same as minsky().model). Wires will automatically be carried through if the addItem method is used.
2) Insertion of all items and groups into the selection. You will need to do this step prior to step 1), because it is not possible for an item to belong to two groups at once, with the exception of the selection.

I think the key is to resolve why the simplest approach of not inserting the temporary group (mentioned below) causes a segfault. Time to break out gdb.

Reply:
1. Thanks. The fact that minsky().model and canvas.model are not necessarily the same has helped me to find a way to perform the pasting operation by looping separately over the two types of items and toggling their selection status. That way only the pasted items remain in selection:

`      auto p = minsky().model;
	for (auto i: p->items)
	    if (!dynamic_cast<IntOp*>(i.get())) canvas.selection.toggleItemMembership(i);   
	p->moveContents(*g);
	auto q = canvas.model;
	for (auto i: q->items)
	    canvas.selection.toggleItemMembership(i);   	`

For some reason an IntOp item that has not been selected also remains in selection if I don't add " if (!dynamic_cast<IntOp*>(i.get()))" when looping over the minsky().model items.

2. Doing the above, I now get a more clearly defined crash too, which has something to do with wires:

minsky: model/group.cc:281: void minsky::GroupItems::adjustWiresGroup(minsky::Wire&): Assertion `p1 && p2' failed.

The error again disappears when I reinstate canvas.setItemFocus(model->addGroup(g)); a few lines up. Very strange. I will try to finish sorting it out when I get back from my trip tomorrow evening and continue working on the other bugs on Steve's list.

Reply:
This also strikes me as a bit bizarre. The usual scenario is that canvas.model == minsky.model. So this will basically be the identity operation. When canvas.mode != minsky,model, it ends up filling the selection with the contents of both groups.

The assertion failure indicates that one of the wire ends is not owned by a group.

Reply:
Will I need to pass the coordinates as arguments to Minsky::paste() to add all the wire ends via  the ports they are attached to? I have tried several approaches to remove canvas.itemFocus(model->addGroup(g)) and p->removeGroup(*g), but the best I could get also deleted the wire during the process...

Reply:
No - the actual wired coordinate data is relative to the end points, which is decided by the port coordinates, which in turn is relative to the coordinates of the items to which they belong.

Group::adjustWires will move relevant wires to the groups that should own them.

Reply:
Thanks. 

I see Group::adjustWires is called in Group::moveContents, so that should be taken care of.